src/test/ui/issues/issue-22638.stderr

```diff
@@ -8,7 +8,7 @@ LL | |         a.matches(f)
 LL | |     }
    | |_____^
    |
-   = note: consider adding a `#![type_length_limit="26214380"]` attribute to your crate
+   = note: consider adding a `#![type_length_limit="30408681"]` attribute to your crate
```

Comment:
@rust-lang/compiler This is a bit silly, can we remove `type_length_limit`?
Does it do anything for us nowadays? Other than slow down compilation.

Reply:
I assume by "remove" you mean "make it do nothing"?

Reply:
Oh, right, it's stable. The other thing we can do is count only unique types, which should make cases like these far less drastic presumably.

Reply:
I'm confused -- why would it not do something for us? There are various cases where the size of types and amount of computational work can balloon without generating large stacks (also true: in the SLG solver Chalk at least, we don't even have a concept of overflow apart from type-size limits, though depending on what precise solver design we wind up with, we could keep something like overflow, but it's not a great concept because it makes results for a given query dependent on context instead of independent).

Reply:
That said, this is an orthogonal place where we could apply type-limits, instead of applying them at type creation time or whatever.

Reply:
I think it was added originally to not generate LLVM type names that were too long or something, which is a silly problem IMO.

At the very least we should switch to a DAG approach (i.e. don't count duplicates), I assume Chalk at least is similarly proportional in unique types, not total occurrences.

Reply:
I'm not sure what duplicates exactly means? Oh, I guess like `Foo<(A, A)>` only counts `A` once? Chalk is currently just counting the total number of types I think. I'd have to think if the DAG version "suffices". I'm not sure, why is that much better? I'm just a bit confused about why you care about this I guess. Is it something that you've seen arise frequently?

Reply:
> Is it something that you've seen arise frequently?

It's almost *guaranteed* to arise with closures in generic functions, the generic shows up once in the parent `Substs` and then in one or more capture types.

That multiplies during monomorphization, and can blow out of proportions, *despite* nothing actually operating proportionally to the multiple occurrences.

Remember #54540? :rofl: